_posts/2015-12-30-Python-Join-Tutorial1.md

```diff
@@ -0,0 +1,302 @@
+
+## Demonstration Data
+
+As previously (and here we mean "versus" the solutions [in java][7], [with hive][8] and [with pig][9]) we use two datasets called `users` and `transactions`. 
+
+{% highlight bash %}
+users
+1	matthew@test.com	EN	US
+2	matthew@test2.com	EN	GB
+3	matthew@test3.com	FR	FR
+{% endhighlight %}
+
+and
+
+{% highlight bash %}
+transactions
+1	1	1	300	a jumper
+2	1	2	300	a jumper
+3	1	2	300	a jumper
+4	2	3	100	a rubber chicken
+5	1	3	300	a jumper
+{% endhighlight %}
+
+**This time though they will constitute the same input dataset!** Soon it will become clear why.
+
+{% highlight bash %}
+hdfs dfs -mkdir input
+
+hdfs dfs -put ./users.txt input
+hdfs dfs -put ./transactions.txt input
+{% endhighlight %}
+
+And for command line testing
+{% highlight bash %}
+cat ./transactions.txt ./users.txt > ./transactions_and_users.txt
+{% endhighlight %}
+
+## Code
+
+We divide the task into two parts. First part solves the question of a join. Second part is responsible for counting distinct by key.
+
+The code for the both parts of the solution and data used in this post can be found in my [`hive examples` GitHub repository][github].
+
+## Join
+
+Mapper:
+
+{% highlight python %}
+#!/usr/bin/env python
+import sys
+for line in sys.stdin:
+	user_id = ""
+	product_id = "-"
+	location = "-"
+	line = line.strip()         
+	splits = line.split("\t")         
+	if len(splits) == 5:
+		user_id = splits[2]
+		product_id = splits[1]
+	else:
+		user_id = splits[0]
+		location = splits[3]                   
+	print '%s\t%s\t%s' % (user_id,product_id,location)
+{% endhighlight %}
+
+Reducer:
+{% highlight python %}
+#!/usr/bin/env python
+import sys
+import string
+
+last_user_id = None
+cur_location = "-"
+
+for line in sys.stdin:
+    line = line.strip()
+    user_id,product_id,location = line.split("\t")
+
+    if not last_user_id or last_user_id != user_id:
+        last_user_id = user_id
+        cur_location = location
+    elif user_id == last_user_id:
+        location = cur_location
+        print '%s\t%s' % (product_id,location)
+{% endhighlight %}
+
+The Mapper reads both datasets and distinguishes them by the number of columns. Transactions data has more columns than the users data.
+
+The aim of the mapper is to select the data we need to calculate the number of unique locations in which each product has been sold, namely user_id, poduct_id and location, and represent the data in such a way so it would be easy for the reducer to read it line by line. The Reducer will substitute an absent value of location into the rows where user_id and product_id are present. That is the rows from the transactions table.
+
+Using the fact that the data will be sorted by key in Hadoop, we sort it by user_id and product_id (using `-Dstream.num.map.output.key.fields=2`) when running in hadoop-streaming, or applying `sort` when testing from command line: 
+
+{% highlight bash%}
+cat transactions_and_users.txt | ./joinMapperTU.py | sort
```

Comment:
For testing here - if we have two input files we can just use wildcards to cat the data:

`cat inputs/*.txt | ./joinMapperTU.py | sort`